### DISCUSSION TITLE
"Error: No matching key" in dapp

### DISCUSSION THREAD

**Discussion by ian-eulith** [WalletConnect/walletconnect-monorepo#3887] [Developer Support]
While working on our wallet implementation, we have found that sometimes sessions get corrupted so that initiating a transaction from a dapp causes a `No matching key, keychain: <hex string>` *in the dapp*. I read [this discussion](https://github.com/orgs/WalletConnect/discussions/3291) but in that case the errors were appearing in the wallet.

Based on the Chrome network logs, our wallet is not receiving anything from the dapp at all, even though our wallet still registers the dapps as connected. Refreshing either page (wallet – our wallet is a webpage – or dapp) does not work. We see the same error logs when attempting to disconnect the wallet on the dapp side. Usually the only way to fix it is by manually clearing local storage in the dapp.

This has happened with multiple different dapps (GMX, Matcha, Curve) seemingly at random. Sometimes when the wallet is connected to multiple dapps, one dapp will be fine and the other will be corrupted.

We can't tell whether the issue is in the dapp(s), our wallet, or the WalletConnect SDK, so we'd appreciate any guidance.

**Comment by ian-eulith** [comment]
We've noticed that `session` and `keychain` are stored separately in local storage and from reading [the code](https://github.com/WalletConnect/walletconnect-monorepo/blob/v2.0/packages/sign-client/src/controllers/engine.ts#L419) it seems possible that the two could get out of sync. We've also noticed that in our wallet, the `wc@2:core:0.3//keychain` slot has many more entries than there are sessions in `wc@2:core:0.3//session`. Might that be related?

**Reply by MuckT** [reply]
@ian-eulith are you still seeing this issue? [Pinning our versions](https://github.com/valora-inc/wallet/pull/4406) to `2.10.0` prior to [this PR](https://github.com/WalletConnect/walletconnect-monorepo/pull/3574) being merged appears to be working.

**Reply by ian-eulith** [reply]
@MuckT thanks for pointing us to that PR. It certainly looks related to the issues we've been seeing. Unfortunately in our case these issues mainly arise on the dapp side, and we don't have any control over what version of the WalletConnect client they are using.
